tools/nbs/filter.py

```diff
@@ -0,0 +1,64 @@
+#!/usr/bin/env python3
+
+import fnmatch
+
+from tqdm import tqdm
+
+from .nbs import Decoder, Encoder
+
+
+def register(command):
+    command.help = (
+        "Filter an nbs stream, removing messages from it. ",
+        "The command will first remove any patterns provided with -r ",
+        "(or everything if no remove patterns are given) ",
+        "and then keep any of those removed that match the -k patterns",
+    )
+
+    # Command arguments
+    command.add_argument("files", metavar="files", nargs="+", help="The nbs files to filter")
+    command.add_argument("-k", "--keep", action="append", help="A message pattern to keep")
+    command.add_argument("-r", "--remove", action="append", help="A message pattern to remove")
```

Comment:
Probably useful to include an example of how to use these? Is it just regex? How do we handle multiple search patterns for `-r` and `-k`? Is it possible to specify `-r` or `-k` more than once?

Reply:
I suppose, and yes you can have as many k/r as you want. It's not regex though, it's the same as shell uses with * and ?. I could go to regex though... thoughts?

Reply:
I think regex would be good 👍, would be the least surprising thing.

Also, by message pattern, do you mean a pattern to match the message name? Can it match nested message names as well? Do those have a `.`?

Reply:
Shell globbing is probably fine, but regex is probably more understood (if people actually understand regex). Either way we need to be careful that the shell doesn't screw up the pattern anyway. 

I honestly don't really see us using too much in the way of wildcards for this, probably more likely that we will be listing all the messages we want/don't want. If we are using wildcards its probably to avoid typing `message.` all the time (`*Image` rather than `message.input.Image`). Is it worth perhaps changing to a comma-separated list for `-k` and `-r`, we can split on the comma and then do regex/globbing on the splits.

Reply:
The main reason I went with shell globbing over regex was that if you did regex, all the dots in message.input.Sensors would technically need to be escaped (although it would probably work regardless and `message\.input\.Sensors` just looks overly complicated

Reply:
`.*Sensors` although that gets you DarwinSensors too